packages/shorebird_cli/lib/src/cache.dart

```diff
@@ -191,38 +238,64 @@ allowed to access $storageUrl.''',
         logger.detail(
           '[cache] optional artifact: "$fileName" was not found, skipping...',
         );
-        return;
+        return response;
       }
 
       throw CacheUpdateFailure(
         '''Failed to download $fileName: ${response.statusCode} ${response.reasonPhrase}''',
       );
     }
 
-    final artifactDirectory = Directory(p.dirname(file.path));
-    await extractArtifact(response.stream, artifactDirectory.path);
+    return response;
+  }
 
-    final expectedChecksum = checksum;
-    if (expectedChecksum != null) {
-      if (!checksumChecker.checkFile(file, expectedChecksum)) {
-        // Delete the artifact directory, so if the download is retried, it will
-        // be re-downloaded.
-        artifactDirectory.deleteSync(recursive: true);
-        throw CacheUpdateFailure(
-          '''Failed to download $fileName: checksum mismatch''',
-        );
-      } else {
-        logger.detail(
-          '''No checksum provided for $fileName, skipping file corruption validation''',
-        );
-      }
+  Future<File> _downloadFile(http.StreamedResponse response) async {
+    final isZip =
+        response.headers[HttpHeaders.contentTypeHeader] == 'application/zip';
+    final tempDirectory = Directory.systemTemp.createTempSync();
+    final tempFile = File(
+      p.join(tempDirectory.path, '$fileName${isZip ? '.zip' : ''}'),
+    );
+    await response.stream.pipe(tempFile.openWrite());
+    return tempFile;
+  }
+
+  void _verifyChecksum({
```

Comment:
This is used for two different kind of verifications it seems?  Are you sure you want to make this function have this implicit sha256 behavior?  Maybe I'm just reacting to using the "expectedChecksum" locals, and what I'm implying is pulling most of this function out into a helper and making these two branches of the if just call the helper with an explicit expectedChecksum/algorithm set?

Reply:
I'm not totally sure I follow. 

> This is used for two different kind of verifications it seems?

We can currently get checksums from two different places (although I added a TODO later on to support always getting the checksum from the header), but otherwise they're used in exactly the same manner (i.e., get the checksum from the file, make sure it matches what we expect it to).

> Are you sure you want to make this function have this implicit sha256 behavior?

By "implicit sha256 behavior", do you mean that we use the hardcoded sha256 hash and ignore the response headers if it's present? 

Maybe this is a naming thing? Would something like `_findAndVerifyChecksum` be more appropriate?

Reply:
I think it's fine.

I think where I got wrapped up is that we have logically two separate methods.  One is "decide what type of verification to do" and the other is "do verification X" and it's currently spread across 2 methods (this one and the ChecksumVerifier below).  I probably would have sharded it slightly differently, but I think I'm mostly tired and this factoring detail wont matter. Happy to leave the doer to decide. :)